### PR TITLE
Concatenates nested namespaces (capability introduced in C++17 std)

### LINKED ISSUE #370: Modernise: Use concatenated nested namespaces instead of hierarchical nested namespaces
C++17 introduced the ability to use concatenated nested namespaces instead of hierarchical nesting of namespace blocks. This is what C++20 & beyond standards uses right within the standard library.

#### References:
- https://stackoverflow.com/a/28022457
- https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4230.html
- https://www.youtube.com/watch?v=-CTPds55l2U
- https://rules.sonarsource.com/cpp/RSPEC-5812

#368 shall close this.

### PR DESCRIPTION
**Pull request by krishnakumarg1984**
C++17 introduced the ability to use concatenated nested namespaces instead of hierarchical nesting of namespace blocks. This is what C++20 & beyond standards uses right within the standard library.

#### References:
- https://stackoverflow.com/a/28022457
- https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4230.html
- https://www.youtube.com/watch?v=-CTPds55l2U
- https://rules.sonarsource.com/cpp/RSPEC-5812

Closes #370.

### CODE REVIEW

File: cpp/sopt/objective_functions.h [hunk 1]
```diff
@@ -31,10 +30,8 @@ std::function<t_real(T)> const unconstrained_l1_regularisation(
     const t_real &gamma, const t_real &sig, const T &y,
     const sopt::LinearTransform<T> &measurement_operator,
     const sopt::LinearTransform<T> &wavelet_operator);
-}  // namespace objective_functions
-}  // namespace sopt
-namespace sopt {
-namespace objective_functions {
+} // namespace sopt::objective_functions
+namespace sopt::objective_functions {
```
Comment (SJaffa) [inline cpp/sopt/objective_functions.h:35]: Why does this close the `sopt::objective_functions` namespace and then immediately repoen it? Would it be better to just remove lines 34-37 so the namespace continues to line 70?

File: cpp/sopt/proximal.h [hunk 2]
```diff
@@ -11,9 +11,7 @@
 #include "sopt/mpi/utilities.h"
 #endif
 
-namespace sopt {
-//! Holds some standard proximals
-namespace proximal {
+namespace sopt::proximal {
```
Comment (SJaffa) [inline cpp/sopt/proximal.h:16]: The comment here doesn't seem very useful but if they put it in I think we should keep it for the documentation.

File: cpp/sopt/proximal_expression.h [hunk 3]
```diff
@@ -6,9 +6,7 @@
 #include <Eigen/Core>
 #include "sopt/maths.h"
 
-namespace sopt {
-//! Holds some standard proximals
-namespace proximal {
+namespace sopt::proximal {
```
Comment (SJaffa) [inline cpp/sopt/proximal_expression.h:11]: Again I think we should keep the comment for docs.

File: cpp/sopt/wavelets/innards.impl.h [hunk 4]
```diff
@@ -5,8 +5,7 @@
 #include <Eigen/Core>
 
 // Function inside anonymouns namespace won't appear in library
-namespace sopt {
-namespace wavelets {
+namespace sopt::wavelets {
 namespace {
```
Comment (SJaffa) [inline cpp/sopt/wavelets/innards.impl.h:9]: Why do we have this unnamed namespace as well? It seems to contain all the tings that are in `sopt::wavelets`, does it serve any function?

File: cpp/sopt/wavelets/sara.cc [hunk 5]
```diff
@@ -1,6 +1,4 @@
 #include "sara.h"
```
Comment (SJaffa) [inline cpp/sopt/wavelets/sara.cc:1]: All this file does is include some headers and create a namespace,there are no functions in that namespace. Why does this file exist?

File: cpp/sopt/credible_region.h [hunk 6]
```diff
@@ -54,11 +53,9 @@ credible_interval(const Eigen::MatrixBase<T> &solution, const t_uint &rows, cons
                   const std::tuple<t_uint, t_uint> &grid_pixel_size,
                   const std::function<t_real(typename T::PlainObject)> &objective_function,
                   const t_real &alpha);
-}  // namespace credible_region
-}  // namespace sopt
+} // namespace sopt::credible_region
```
Comment (SJaffa) [inline cpp/sopt/credible_region.h:57]: This seems to be another place where we close a namespace and immediately open it again. Any ideas why?

File: cpp/sopt/wavelets/innards.impl.h (hunk 4 above)
Comment (SJaffa) [reply cpp/sopt/wavelets/innards.impl.h:9]: Ahh I've just seen the comment above that explains why they did this. Is this a good/common practice?

File: cpp/sopt/wavelets/sara.cc (hunk 5 above)
Comment (krishnakumarg1984) [reply cpp/sopt/wavelets/sara.cc:1]: This happens in a few other files too. I think we should just merge them up (if we manage to hunt them). 

But perhaps I should do that in a separate PR (with a tag of `code=quality`)?

File: cpp/sopt/credible_region.h (hunk 6 above)
Comment (krishnakumarg1984) [reply cpp/sopt/credible_region.h:57]: Yes. As you observed, this happens in a couple of other files too. We may clean this up by merging these adjacent repeated namespace blocks.

We may designate this current PR just for 1:1 verbatim replacements on the current codebase. We can submit this cleanup (and other remaining fixes) in a separate/smaller code-quality PR. Please let me know what you'd prefer.

File: cpp/sopt/wavelets/innards.impl.h (hunk 4 above)
Comment (krishnakumarg1984) [reply cpp/sopt/wavelets/innards.impl.h:9]: Yes. It is a reasonable practice. As per learncpp, https://www.learncpp.com/cpp-tutorial/unnamed-and-inline-namespaces/

>The other effect of unnamed namespaces is that all identifiers inside an unnamed namespace are treated as if they had internal linkage, which means that the content of an unnamed namespace can’t be seen outside of the file in which the unnamed namespace is defined.
> Unnamed namespaces are typically used when you have a lot of content that you want to ensure stays local to a given file, as it’s easier to cluster such content in an unnamed namespace than individually mark all declarations as static. Unnamed namespaces will also keep user-defined types (something we’ll discuss in a later lesson) local to the file, something for which there is no alternative equivalent mechanism to do.

File: cpp/sopt/proximal_expression.h (hunk 3 above)
Comment (krishnakumarg1984) [reply cpp/sopt/proximal_expression.h:11]: Yes. Agreed. Will fix and update the PR. Thank you.

File: cpp/sopt/proximal.h (hunk 2 above)
Comment (krishnakumarg1984) [reply cpp/sopt/proximal.h:16]: Yes. Agreed. Will fix and update the PR. Thank you.

File: cpp/sopt/objective_functions.h (hunk 1 above)
Comment (krishnakumarg1984) [reply cpp/sopt/objective_functions.h:35]: I found this exact thing in one or two files. But I thought this PR should solely be about a 1:1 replacement of the current codebase.

I agree that closing and reopening a namespace immediately isn't particularly useful, and I am happy to roll up this and other leftover code-quality issues into a smaller, separate PR. Please let me know your thoughts.

File: cpp/sopt/proximal.h (hunk 2 above)
Comment (krishnakumarg1984) [reply cpp/sopt/proximal.h:16]: Fixed. Thanks a lot for pointing this out.

File: cpp/sopt/proximal_expression.h (hunk 3 above)
Comment (krishnakumarg1984) [reply cpp/sopt/proximal_expression.h:11]: Thank you for pointing this out. Fixed.

File: cpp/sopt/wavelets/sara.cc (hunk 5 above)
Comment (krishnakumarg1984) [reply cpp/sopt/wavelets/sara.cc:1]: @SJaffa I looked into this. It appears that this file is used by [`cpp/sopt/CMakeLists.txt` in line 26](https://github.com/astro-informatics/sopt/blob/203106724c9b05e7a5d362a21349ef1864f42301/cpp/sopt/CMakeLists.txt#L26), where it says:

```cmake
list(APPEND sources mpi/communicator.cc mpi/registered_types.cc mpi/session.cc wavelets/sara.cc)
```

I deleted this file `cpp/sopt/wavelets/sara.cc` and removed this source from the relevant line in  `cpp/sopt/CMakeLists.txt`, and it builds fine and all tests pass (locally). 

We may designate this current PR just for 1:1 verbatim replacements on the current codebase. We can submit this cleanup (and other remaining fixes like opening and closing namespaces immediately within the same file) in a separate/smaller `code-quality` PR. Please let me know what you'd prefer.

File: cpp/sopt/objective_functions.h (hunk 1 above)
Comment (SJaffa) [reply cpp/sopt/objective_functions.h:35]: Yes I agree that having this one just be a 1:1 switch and make other changes separately sounds good.

**Review by SJaffa — APPROVED** [review]

File: cpp/sopt/objective_functions.h (hunk 1 above)
Comment (krishnakumarg1984) [reply cpp/sopt/objective_functions.h:35]: Recorded in #390 and should be fixed by #392.

File: cpp/sopt/wavelets/sara.cc (hunk 5 above)
Comment (krishnakumarg1984) [reply cpp/sopt/wavelets/sara.cc:1]: Thank you. Have recorded this in #393.  Should be closed by #394.

File: cpp/sopt/credible_region.h (hunk 6 above)
Comment (krishnakumarg1984) [reply cpp/sopt/credible_region.h:57]: Thank you. Recorded in issue #390, which should be closed by PR #392.

**Pull request merged**